Adds lazyRender support

Duder-onomy opened the pull request:
This PR adds support to defer the render of the `yield` block in the `ember-attacher-inner` component. 

The motivation would be to defer the initialization work of the attached items until the user interacts with it. In cases where the yielded contents are complex, this is preferred over having many items initialize which might or might not ever be used. 

Adds a new property called `lazyRender` to the parent `ember-attacher` component. 
This `lazyRender` property defaults to false to maintain current expected behavior. 
Adds tests for lazyRendering to ensure contents are not rendered until after interacting.
Adds working example to the dummy app. 

@kybishop, I am looking for some feedback. Are there any additional tests or cases you want explored? 
I believe that once merged you will also need to deploy the dummy app to github pages. 
Also, I was not sure if this should be in the readme or not as many of the options are not currently enumerated there. 

Closes #74
Closes #64 (in my opinion)

kybishop commented:
Thanks for the PR @Duder-onomy! I'll have some time to look at in depth tonight and tomorrow.

kybishop commented on addon/components/ember-attacher-inner.js:
```diff
@@ -18,6 +18,8 @@ export default Component.extend({
   init() {
     this._super(...arguments);
 
+    this.set('_shouldRenderYieldBlock', !this.get('lazyRender'));
```
`_shouldRenderYieldBlock` could be moved into a computed that depends on 'lazyRender'. Also, I think we should call it `_shouldRender` since it also takes care of rendering the arrow and circle (for the fill animation).

```javascript
_shouldRender: computed('lazyRender', function() {
  return !this.get('lazyRender');
})
```

Setting this property on our first show event should still work as expected.

Side note: As a matter of best practices, use `this.myProperty = something` in `init` since `this.set()` is used to notify ember of property changes. Since we're initializing the component, there is no need to notify of changes. Basically comes down to a small performance win by avoiding unesc work.

kybishop commented on addon/components/ember-attacher.js:
```diff
@@ -52,6 +52,7 @@ export default Component.extend({
   showDuration: DEFAULTS.showDuration,
   showOn: DEFAULTS.showOn,
   target: null,
+  lazyRender: DEFAULTS.lazyRender,
```
I'd love to keep this list alphabetically sorted

kybishop commented on addon/defaults.js:
```diff
@@ -16,5 +16,6 @@ export default {
   showOn: 'mouseenter focus',
-  tooltipClass: 'ember-attacher-popper ember-attacher-tooltip'
+  tooltipClass: 'ember-attacher-popper ember-attacher-tooltip',
+  lazyRender: false
```
I'd love to keep this list alphabetically sorted

kybishop commented on addon/templates/components/ember-attacher.hbs:
```diff
@@ -7,7 +7,7 @@
                 popperContainer=popperContainer
                 renderInPlace=renderInPlace
-                target=target as |emberPopper|}}
+                target=target as |emberPopper|}}                      
```
Looks like some extra whitespace snuck in here

kybishop commented on tests/dummy/app/services/popover-data.js:
```diff
@@ -13,5 +13,6 @@ export default Service.extend({
   showOn: 'click',
-  target: '.target-plz'
+  target: '.target-plz',
+  lazyRender: false
```
I'd love to keep this list alphabetically sorted

kybishop commented on tests/dummy/app/templates/components/attachment-example.hbs:
```diff
@@ -40,6 +40,7 @@
                       showDelay=popoverData.showDelay
                       showDuration=popoverData.showDuration
                       showOn=popoverData.showOn
+                      lazyRender=popoverData.lazyRender
```
alphabetically sorted please :)

kybishop commented on tests/dummy/app/templates/components/attachment-example.hbs:
```diff
@@ -304,6 +305,22 @@
                           {{/power-select}}"
       </centered>
     </hbox>
+    {{#unless isConfiguringTooltip}}
```
Is there a particular reason to hide this when people are playing with `attach-tooltip`?

Also, let's keep this alphabetically sorted within the list of configurables :)

Duder-onomy commented:
Thanks for the feedback @kybishop. I think I addressed everything you mentioned and added a second test to ensure the default current behavior when lazyRender is unset.

Looking forward to getting this in. Been enjoying using `ember-attacher` recently, having `ember-popper` available on the yielded hash is a clutch move BTW. Saved me once already.

kybishop commented:
Merged! Thanks a ton @Duder-onomy. I'll have this out with the next release, which will be before EOW

Duder-onomy commented:
RAWK!
